devito/archinfo.py

```diff
@@ -60,18 +62,36 @@ def get_cpu_brand():
             except:
                 pass
 
-            return None
+            try:
+                return cpuinfo.get_cpu_info().get('raw_arch_string')
+            except:
+                pass
 
         cpu_info['flags'] = get_cpu_flags()
         cpu_info['brand'] = get_cpu_brand()
 
-    if not all(i in cpu_info for i in ('flags', 'brand')):
-        # Fallback
-        ci = cpuinfo.get_cpu_info()
-        cpu_info['flags'] = ci.get('flags')
-        cpu_info['brand'] = ci.get('brand')
+    try:
+        if 'flags' not in cpu_info:
+            # Fallback
+            ci = cpuinfo.get_cpu_info()
+            cpu_info['flags'] = ci.get('flags')
+    except:
+        pass
 
     # Detect number of logical cores
+    try:
+        if cpu_info['brand'] == 'aarch64':
+            # In some ARM processors psutils and lscpu fail to detect cores correctly
+            logical = psutil.cpu_count(logical=True)
+            physical = psutil.cpu_count(logical=False)
+            if physical != (lscpu()['Core(s) per socket'] * lscpu()['Socket(s)']):
+                physical = (lscpu()['Core(s) per socket'] * lscpu()['Socket(s)'])
+            cpu_info['logical'] = logical
+            cpu_info['physical'] = physical
+            return cpu_info
```

Comment:
We can rewrite this try-except as:

```
if cpu_info['brand'] == 'aarch64':
    # Special case: ...
    cpu_info['physical'] = lscpu()['Core(s) per socket'] * lscpu()['Socket(s)']
```

we don't need the `logical`; we can move it right below lines 95-104, which will still work. 

In particular, this should go below the comment "Detect number of physical cores"

Reply:
but then we will not be able to return cpu_info for the case of arm?

Reply:
note: `if cpu_info['brand']` fails on OSX, cpu_info does not have 'brand'